cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -12,113 +12,150 @@
 package processes
 
 import (
-	"bufio"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"os"
-	"path"
+	"path/filepath"
 	"strings"
 	"time"
 
+	"github.com/palantir/stacktrace"
 	"golang.org/x/crypto/ssh"
 
 	"alluxio.org/cli/env"
 	"alluxio.org/log"
 )
 
-func getMasters() ([]string, error) {
-	mastersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "masters")
-	mastersFile, err := os.Open(mastersDir)
-	if err != nil {
-		log.Logger.Errorf("Error reading worker hostnames at %s", mastersDir)
-		return nil, err
+var cliPath = filepath.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+var privateKeySigner ssh.Signer
+var masterList []string
+var workerList []string
+
+// for each master, create a client and run
+// TODO: now start master one by one, need to do them in parallel
+func runCommand(command string, onMasters bool, onWorkers bool) error {
+	// get list of masters and workers
+	var nodes []string
+	if onMasters {
+		if len(masterList) == 0 {
+			if err := getNodes(true); err != nil {
+				log.Logger.Fatalf("Cannot get masters, error: %s", err)
+			}
+		}
+		nodes = append(nodes, masterList...)
+	}
+	if onWorkers {
+		if len(workerList) == 0 {
+			if err := getNodes(false); err != nil {
+				log.Logger.Fatalf("Cannot get workers, error: %s", err)
+			}
+		}
+		nodes = append(nodes, workerList...)
+	}
+
+	// get public key if nones
+	if privateKeySigner == nil {
+		if err := getPrivateKey(); err != nil {
+			log.Logger.Fatalf("Cannot get private key, error: %s", err)
+		}
 	}
 
-	mastersReader := bufio.NewReader(mastersFile)
-	var mastersList []string
-	lastLine := false
-	for !lastLine {
-		// read lines of the workers file
-		line, err := mastersReader.ReadString('\n')
+	// dial nodes, run sessions and close
+	var errors []error
+	for _, node := range nodes {
+		conn, err := dialConnection(node, privateKeySigner)
 		if err != nil {
-			if err == io.EOF {
-				lastLine = true
-			} else {
-				log.Logger.Errorf("Error parsing worker file at this line: %s", line)
-				return nil, err
-			}
+			errors = append(errors, err)
+			continue
 		}
-		// remove notes
-		if strings.Index(line, "#") != -1 {
-			line = line[:strings.Index(line, "#")]
+		if err := runSession(node, conn, command); err != nil {
+			errors = append(errors, err)
```

Comment:
the current logic will not exit with an error before the connection is closed but it does not follow the standard pattern of using `defer` to close it. this can be dangerous in the future because if a new command needs to be added in here, it also needs to be careful to append an error to the list and avoid running if the previous command is an error. this can quickly become messy. a more conventional way to express this is:
```
var errs []error
for _, node := range nodes {
  if err := runCommandOnNode(); err != nil {
    errs = append(errs, stacktrace.Propagate(err, "error running command on node %v", node)
  }
}
```
with a separate function
```
func runCommandOnNode() error {
  conn, err := dialConnection()
  if err != nil {
    return stacktrace.Propagate(err)
  }
  defer conn.Close()
  if err := runCommand(); err != nil {
    return stacktrace.Propagate(err)
  }
  return nil
}
```

this is such a common pattern for concurrent execution that there is a built in struct to handle this situation: see `sync.Errgroup`